src/smt/theory_str_noodler/decision_procedure.cpp

```diff
@@ -0,0 +1,301 @@
+                            std::vector<BasicTerm> new_inclusion_left_side{ left_var };
+                            const auto &new_inclusion = new_element.inclusion_graph->add_node(new_inclusion_left_side, left_side_vars_to_new_vars[i]);
+                            // we also add this inclusion to the worklist, as it represents unification
+                            // we push it to the front if we are processing node that is not on the cycle, because it should not get stuck in the cycle then
+                            // TODO: is this correct? can we push to the front?
+                            new_element.push_unique(new_inclusion, is_node_to_process_on_cycle);
+                        } else {
+                            // TODO make this function or something, we do the same thing here as for the right side when substituting
+                            // left_var is not substitued by anything yet, we will substitute it
+                            substitution_map[left_var] = left_side_vars_to_new_vars[i];
+                            // as left_var wil be substituted in the inclusion graph, we do not need to remember the automaton assignment for it
+                            new_element.aut_ass.erase(left_var);
+                            // update the length variables
+                            // TODO: is this enough to update variables only when substituting?
+                            if (new_element.length_sensitive_vars.count(left_var) > 0) { // if left_var is length-aware => substituted vars should become length-aware
+                                for (const BasicTerm &new_var : left_side_vars_to_new_vars[i]) {
+                                    new_element.length_sensitive_vars.insert(new_var);
+                                }
+                            }
+                        }
+                    }
+
+                    // do substitution in the inclusion graph
+                    new_element.substitute_vars(substitution_map);
+
+                    // update inclusion graph edges
+                    new_element.inclusion_graph->add_inclusion_graph_edges();
+
+                    // update the substitution_map of new_element by the new substitutions
+                    new_element.substitution_map.merge(substitution_map);
+
+                    // TODO should we really push to front when not on cycle?
+                    if (!is_node_to_process_on_cycle) {
+                        worklist.push_front(new_element);
+                    } else {
+                        worklist.push_back(new_element);
+                    }
+
+                    // TODO: should we do something more here??
+
+                }
+                
+                ++noodlification_no; // TODO: when to do this increment??
+            }
+        }
+
+        return false;
+    }
```

Comment:
Is it possible to split this huge method into sevaral smaller ones? So far it is really difficult to read.